File: src/main/java/com/enthusia/koth/application/rules/RestrictionService.java
package com.enthusia.koth.application.rules;

import com.enthusia.koth.domain.event.ActiveEvent;
import com.enthusia.koth.domain.rules.RestrictedItemType;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import java.time.Duration;
import java.time.Instant;
import java.util.EnumMap;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

public final class RestrictionService {
    private static final Set<Material> SPEARS = Set.of(
            Material.WOODEN_SPEAR,
            Material.STONE_SPEAR,
            Material.COPPER_SPEAR,
            Material.IRON_SPEAR,
            Material.GOLDEN_SPEAR,
            Material.DIAMOND_SPEAR,
            Material.NETHERITE_SPEAR
    );

    private final Map<UUID, Map<RestrictedItemType, Instant>> cooldowns = new ConcurrentHashMap<>();

    public RestrictionDecision canUseItem(Player player, ActiveEvent event, ItemStack item) {
        if (item == null) {
            return RestrictionDecision.allowed(Duration.ZERO);
        }
        RestrictedItemType type = typeFor(item.getType());
        if (type == null) {
            return RestrictionDecision.allowed(Duration.ZERO);
        }
        return decision(player, event, type);
    }

    public RestrictionDecision canDealDamageWith(Player player, ActiveEvent event, ItemStack item) {
        if (item == null) {
            return RestrictionDecision.allowed(Duration.ZERO);
        }
        Material material = item.getType();
        if (material == Material.MACE) {
            return switch (event.request().rules().maceRule()) {
                case FULLY_DISABLED -> RestrictionDecision.denied("Mace damage is disabled for this KOTH.");
                case BREACH_DISABLED -> RestrictionDecision.allowed(event.request().rules().maceCooldown());
                case DENSITY_DISABLED -> RestrictionDecision.allowed(event.request().rules().maceCooldown());
                case FULLY_ALLOWED -> decision(player, event, RestrictedItemType.MACE);
            };
        }
        if (SPEARS.contains(material)) {
            return decision(player, event, RestrictedItemType.SPEAR);
        }
        return RestrictionDecision.allowed(Duration.ZERO);
    }

    public RestrictionDecision canUseElytra(Player player, ActiveEvent event) {
        return decision(player, event, RestrictedItemType.ELYTRA);
    }

    public void applyCooldown(Player player, RestrictedItemType type) {
        cooldowns.computeIfAbsent(player.getUniqueId(), ignored -> new EnumMap<>(RestrictedItemType.class))
                .put(type, Instant.now());
    }

    public void clear() {
        cooldowns.clear();
    }

    public String describeRestriction(ActiveEvent event, RestrictedItemType type) {
        return switch (type) {
            case ELYTRA -> event.request().rules().elytraAllowed() ? "Elytra allowed" : "Elytra disabled";
            case MACE -> "Mace: " + event.request().rules().maceRule();
            case SPEAR -> event.request().rules().spearAllowed() ? "Spear allowed" : "Spear disabled";
            case ENDER_PEARL -> event.request().rules().enderPearlAllowed() ? "Pearls allowed" : "Pearls disabled";
            case WIND_CHARGE -> event.request().rules().windChargeAllowed() ? "Wind charge allowed" : "Wind charge disabled";
        };
    }

    private RestrictionDecision decision(Player player, ActiveEvent event, RestrictedItemType type) {
        Duration cooldown = cooldownFor(event, type);
        if (!enabled(event, type)) {
            return RestrictionDecision.denied(type.name().toLowerCase().replace('_', ' ') + " is disabled for this KOTH.");
        }
        Duration remaining = remainingCooldown(player, type, cooldown);
        if (!remaining.isZero() && !remaining.isNegative()) {
            return RestrictionDecision.denied("Still on cooldown for " + remaining.toSeconds() + "s.");
        }
        return RestrictionDecision.allowed(cooldown);
    }

    private boolean enabled(ActiveEvent event, RestrictedItemType type) {
        return switch (type) {
            case ELYTRA -> event.request().rules().elytraAllowed();
            case MACE -> event.request().rules().maceRule() != com.enthusia.koth.domain.MaceRule.FULLY_DISABLED;
            case SPEAR -> event.request().rules().spearAllowed();
            case ENDER_PEARL -> event.request().rules().enderPearlAllowed();
            case WIND_CHARGE -> event.request().rules().windChargeAllowed();
        };
    }

    private Duration cooldownFor(ActiveEvent event, RestrictedItemType type) {
        return switch (type) {
            case MACE -> event.request().rules().maceCooldown();
            case SPEAR -> event.request().rules().spearCooldown();
            case ENDER_PEARL -> event.request().rules().enderPearlCooldown();
            case WIND_CHARGE -> event.request().rules().windChargeCooldown();
            case ELYTRA -> Duration.ZERO;
        };
    }

    private Duration remainingCooldown(Player player, RestrictedItemType type, Duration cooldown) {
        if (cooldown.isZero() || cooldown.isNegative()) {
            return Duration.ZERO;
        }
        Instant last = cooldowns.getOrDefault(player.getUniqueId(), Map.of()).get(type);
        if (last == null) {
            return Duration.ZERO;
        }
        return cooldown.minus(Duration.between(last, Instant.now()));
    }

    private RestrictedItemType typeFor(Material material) {
        if (material == Material.ELYTRA) return RestrictedItemType.ELYTRA;
        if (material == Material.MACE) return RestrictedItemType.MACE;
        if (SPEARS.contains(material)) return RestrictedItemType.SPEAR;
        if (material == Material.ENDER_PEARL) return RestrictedItemType.ENDER_PEARL;
        if (material == Material.WIND_CHARGE) return RestrictedItemType.WIND_CHARGE;
        return null;
    }
}

File: src/main/kotlin/net/badgersmc/ek/infrastructure/restriction/RestrictionService.kt
package net.badgersmc.ek.infrastructure.restriction

import net.badgersmc.ek.domain.KothEvent
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.time.Clock
import java.time.Duration
import java.time.Instant
import java.util.EnumMap
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/** Types of items that can be restricted per arena. */
enum class RestrictedItemType {
    ELYTRA, MACE, SPEAR, ENDER_PEARL, WIND_CHARGE
}

/** Per-arena mace policy. */
enum class MaceRule {
    FULLY_DISABLED, BREACH_DISABLED, DENSITY_DISABLED, FULLY_ALLOWED
}

data class RuleSet(
    val elytraAllowed: Boolean = true,
    val maceRule: MaceRule = MaceRule.FULLY_ALLOWED,
    val spearAllowed: Boolean = true,
    val enderPearlAllowed: Boolean = true,
    val windChargeAllowed: Boolean = true,
    val maceCooldownSeconds: Int = 0,
    val spearCooldownSeconds: Int = 0,
    val enderPearlCooldownSeconds: Int = 0,
    val windChargeCooldownSeconds: Int = 0,
) {
    companion object {
        val PERMISSIVE = RuleSet()
    }
}

data class RestrictionDecision(
    val allowed: Boolean,
    val message: String,
    val cooldownSeconds: Int,
) {
    companion object {
        fun allowed(cooldownSeconds: Int = 0) = RestrictionDecision(true, "", cooldownSeconds)
        fun denied(message: String) = RestrictionDecision(false, message, 0)
    }
}

/** Immutable launch-time classification used when a projectile later enters an arena. */
data class ProjectileUseSnapshot(
    val eventId: UUID,
    val projectileId: UUID,
    val shooterId: UUID,
    val item: ItemStack?,
    val launchedInsideZone: Boolean,
)

/**
 * Evaluates restrictions and owns all event-scoped transient state.
 *
 * Cooldowns and projectile launch snapshots are keyed by event id. A later event can therefore
 * never inherit state even if a lifecycle callback is missed, while [clearEvent] still releases
 * references immediately on every normal termination path.
 */
class RestrictionService(
    private val rulesForArena: (arenaId: String) -> RuleSet,
    private val clock: Clock = Clock.systemUTC(),
) {
    private data class EventState(
        val cooldowns: MutableMap<UUID, MutableMap<RestrictedItemType, Instant>> = ConcurrentHashMap(),
        val projectiles: MutableMap<UUID, ProjectileUseSnapshot> = ConcurrentHashMap(),
    )

    private val states = ConcurrentHashMap<UUID, EventState>()

    fun canUseItem(player: Player, event: KothEvent, item: ItemStack?): RestrictionDecision {
        val type = item?.type?.let(::typeFor) ?: return RestrictionDecision.allowed()
        return decision(player, event, type)
    }

    fun canUseType(player: Player, event: KothEvent, type: RestrictedItemType): RestrictionDecision =
        decision(player, event, type)

    fun canUseTypeIgnoringCooldown(event: KothEvent, type: RestrictedItemType): RestrictionDecision {
        val rules = rulesForArena(event.arena.id)
        return if (enabled(rules, type)) {
            RestrictionDecision.allowed(cooldownFor(rules, type))
        } else {
            RestrictionDecision.denied("${type.name.lowercase().replace('_', ' ')} is disabled for this KOTH.")
        }
    }

    fun canDealDamageWith(player: Player, event: KothEvent, item: ItemStack?): RestrictionDecision {
        if (item == null) return RestrictionDecision.allowed()
        val rules = rulesForArena(event.arena.id)
        return when (item.type) {
            Material.MACE -> when (rules.maceRule) {
                MaceRule.FULLY_DISABLED -> RestrictionDecision.denied("Mace damage is disabled for this KOTH.")
                MaceRule.BREACH_DISABLED -> if (item.containsEnchantment(Enchantment.BREACH)) {
                    RestrictionDecision.denied("The Breach enchantment is disabled for this KOTH.")
                } else {
                    decision(player, event, RestrictedItemType.MACE)
                }
                MaceRule.DENSITY_DISABLED -> if (item.containsEnchantment(Enchantment.DENSITY)) {
                    RestrictionDecision.denied("The Density enchantment is disabled for this KOTH.")
                } else {
                    decision(player, event, RestrictedItemType.MACE)
                }
                MaceRule.FULLY_ALLOWED -> decision(player, event, RestrictedItemType.MACE)
            }
            else -> if (item.type in SPEARS) {
                decision(player, event, RestrictedItemType.SPEAR)
            } else {
                RestrictionDecision.allowed()
            }
        }
    }

    fun canUseElytra(player: Player, event: KothEvent): RestrictionDecision =
        decision(player, event, RestrictedItemType.ELYTRA)

    fun applyCooldown(player: Player, event: KothEvent, type: RestrictedItemType) {
        states.computeIfAbsent(event.id) { EventState() }
            .cooldowns
            .computeIfAbsent(player.uniqueId) { EnumMap(RestrictedItemType::class.java) }[type] = clock.instant()
    }

    fun recordProjectile(
        event: KothEvent,
        projectileId: UUID,
        player: Player,
        item: ItemStack?,
        launchedInsideZone: Boolean = false,
    ) {
        val snapshot = ProjectileUseSnapshot(
            eventId = event.id,
            projectileId = projectileId,
            shooterId = player.uniqueId,
            item = item?.clone(),
            launchedInsideZone = launchedInsideZone,
        )
        states.computeIfAbsent(event.id) { EventState() }.projectiles[projectileId] = snapshot
    }

    fun projectileSnapshot(event: KothEvent, projectileId: UUID): ProjectileUseSnapshot? =
        states[event.id]?.projectiles?.get(projectileId)?.takeIf { it.eventId == event.id }

    fun removeProjectile(eventId: UUID, projectileId: UUID) {
        states[eventId]?.projectiles?.remove(projectileId)
    }

    fun clearEvent(eventId: UUID) {
        states.remove(eventId)
    }

    fun clear() {
        states.clear()
    }

    internal fun trackedProjectileCount(eventId: UUID): Int = states[eventId]?.projectiles?.size ?: 0

    private fun decision(player: Player, event: KothEvent, type: RestrictedItemType): RestrictionDecision {
        val rules = rulesForArena(event.arena.id)
        val cooldownSeconds = cooldownFor(rules, type)
        if (!enabled(rules, type)) {
            return RestrictionDecision.denied("${type.name.lowercase().replace('_', ' ')} is disabled for this KOTH.")
        }
        val remaining = remainingCooldown(player, event, type, cooldownSeconds)
        return if (remaining > 0) {
            RestrictionDecision.denied("Still on cooldown for ${remaining}s.")
        } else {
            RestrictionDecision.allowed(cooldownSeconds)
        }
    }

    private fun enabled(rules: RuleSet, type: RestrictedItemType): Boolean = when (type) {
        RestrictedItemType.ELYTRA -> rules.elytraAllowed
        RestrictedItemType.MACE -> rules.maceRule != MaceRule.FULLY_DISABLED
        RestrictedItemType.SPEAR -> rules.spearAllowed
        RestrictedItemType.ENDER_PEARL -> rules.enderPearlAllowed
        RestrictedItemType.WIND_CHARGE -> rules.windChargeAllowed
    }

    private fun cooldownFor(rules: RuleSet, type: RestrictedItemType): Int = when (type) {
        RestrictedItemType.MACE -> rules.maceCooldownSeconds
        RestrictedItemType.SPEAR -> rules.spearCooldownSeconds
        RestrictedItemType.ENDER_PEARL -> rules.enderPearlCooldownSeconds
        RestrictedItemType.WIND_CHARGE -> rules.windChargeCooldownSeconds
        RestrictedItemType.ELYTRA -> 0
    }

    private fun remainingCooldown(
        player: Player,
        event: KothEvent,
        type: RestrictedItemType,
        cooldownSeconds: Int,
    ): Int {
        if (cooldownSeconds <= 0) return 0
        val last = states[event.id]?.cooldowns?.get(player.uniqueId)?.get(type) ?: return 0
        val elapsed = Duration.between(last, clock.instant()).seconds
        return (cooldownSeconds - elapsed).coerceAtLeast(0).toInt()
    }

    companion object {
        private val SPEARS: Set<Material> = listOf(
            "WOODEN_SPEAR", "STONE_SPEAR", "COPPER_SPEAR", "IRON_SPEAR",
            "GOLDEN_SPEAR", "DIAMOND_SPEAR", "NETHERITE_SPEAR", "SPEAR",
        ).mapNotNull { Material.getMaterial(it) }.toSet()

        fun typeFor(material: Material): RestrictedItemType? = when (material) {
            Material.ELYTRA -> RestrictedItemType.ELYTRA
            Material.MACE -> RestrictedItemType.MACE
            Material.ENDER_PEARL -> RestrictedItemType.ENDER_PEARL
            Material.WIND_CHARGE -> RestrictedItemType.WIND_CHARGE
            else -> if (material in SPEARS) RestrictedItemType.SPEAR else null
        }
    }
}
